src/util/heap/layout/vm_layout_constants.rs

```diff
@@ -120,48 +34,175 @@ pub const LOG_PAGES_IN_SPACE64: usize = LOG_SPACE_SIZE_64 - LOG_BYTES_IN_PAGE as
+    pub fn force_use_contiguous_spaces(&self) -> bool {
+        self.log_address_space > 35
+    }
+}
 
-/**
- * Bitwise mask to isolate a space index in a virtual address.
- *
- * We can't express this constant in a 32-bit environment, hence the
- * conditional definition.
- */
-#[cfg(target_pointer_width = "32")]
-pub const SPACE_MASK_64: usize = 0;
-#[cfg(target_pointer_width = "64")]
-pub const SPACE_MASK_64: usize = ((1 << LOG_MAX_SPACES) - 1) << SPACE_SHIFT_64;
-
-/*
- * Size of each space in the 64-bit memory layout
- *
- * We can't express this constant in a 32-bit environment, hence the
- * conditional definition.
- */
-// FIXME: When Compiling for 32 bits this expression makes no sense
-// #[allow(const_err)]
-// pub const SPACE_SIZE_64: usize = if_then_else_usize!(HEAP_LAYOUT_64BIT,
-//    1 << LOG_SPACE_SIZE_64, MAX_SPACE_EXTENT);
-#[cfg(target_pointer_width = "64")]
-pub const SPACE_SIZE_64: usize = 1 << LOG_SPACE_SIZE_64;
-#[cfg(target_pointer_width = "32")]
-pub const SPACE_SIZE_64: usize = MAX_SPACE_EXTENT;
+impl VMLayoutConstants {
+    /// Normal 32-bit configuration
+    pub const fn new_32bit() -> Self {
+        Self {
+            log_address_space: 32,
+            heap_start: chunk_align_down(unsafe { Address::from_usize(0x8000_0000) }),
+            heap_end: chunk_align_up(unsafe { Address::from_usize(0xd000_0000) }),
+            vm_space_size: chunk_align_up(unsafe { Address::from_usize(0xdc0_0000) }).as_usize(),
+            log_max_chunks: Self::LOG_ARCH_ADDRESS_SPACE - LOG_BYTES_IN_CHUNK,
+            log_space_extent: 31,
+            space_shift_64: 0,
+            space_mask_64: 0,
+            space_size_64: 1 << 31,
+        }
+    }
+    /// Normal 64-bit configuration
+    #[cfg(target_pointer_width = "32")]
+    pub fn new_64bit() -> Self {
+        unimplemented!("64-bit heap constants do not work with 32-bit builds")
+    }
+    #[cfg(target_pointer_width = "64")]
+    pub fn new_64bit() -> Self {
+        Self {
+            log_address_space: 47,
+            heap_start: chunk_align_down(unsafe {
+                Address::from_usize(0x0000_0200_0000_0000usize)
+            }),
+            heap_end: chunk_align_up(unsafe { Address::from_usize(0x0000_2200_0000_0000usize) }),
+            vm_space_size: chunk_align_up(unsafe { Address::from_usize(0xdc0_0000) }).as_usize(),
+            log_max_chunks: Self::LOG_ARCH_ADDRESS_SPACE - LOG_BYTES_IN_CHUNK,
+            log_space_extent: 41,
+            space_shift_64: 41,
+            space_mask_64: ((1 << 4) - 1) << 41,
+            space_size_64: 1 << 41,
+        }
+    }
+    /// 64-bit configuration with compressed pointers
+    pub fn new_64bit_with_pointer_compression(heap_size: usize) -> Self {
+        assert!(
+            heap_size <= (32usize << LOG_BYTES_IN_GBYTE),
+            "Heap size is larger than 32 GB"
+        );
+        let start = 0x4000_0000;
```

Comment:
Ideally we allow a starting address to be configured instead of a hardcoded starting address

Reply:
Right. I also hardcoded a few OpenJDK-specific heap range optimizations here.

I moved the code to openjdk binding. Bindings can now call `VMLayoutConstants::set_custom_vm_layout_constants` to configure the values to anything they want.